lib/libxdp/libxdp_internal.h

```diff
@@ -62,4 +66,54 @@ LIBXDP_HIDE_SYMBOL struct xdp_program *xdp_program__clone(struct xdp_program *pr
 
 #define ARRAY_SIZE(x) (sizeof(x) / sizeof(*(x)))
 
+/* OPTS macros, from libbpf_internal.h */
+
+static inline bool libbpf_is_mem_zeroed(const char *p, ssize_t len)
+{
+	while (len > 0) {
+		if (*p)
+			return false;
+		p++;
+		len--;
+	}
+	return true;
+}
+
+static inline bool libbpf_validate_opts(const char *opts,
+					size_t opts_sz, size_t user_sz,
+					const char *type_name)
+{
+	if (user_sz < sizeof(size_t)) {
+		pr_warn("%s size (%zu) is too small\n", type_name, user_sz);
+		return false;
+	}
+	if (!libbpf_is_mem_zeroed(opts + opts_sz, (ssize_t)user_sz - opts_sz)) {
+		pr_warn("%s has non-zero extra bytes\n", type_name);
+		return false;
+	}
+	return true;
+}
+
+#define OPTS_VALID(opts, type)						      \
```

Comment:
What about `DECLARE_LIBBPF_OPTS()`? We should probably define a `DECLARE_LIBXDP_OPTS()` macro just for consistency...